networking_aci/extensions/acioperations.py

```diff
@@ -0,0 +1,219 @@
+                                         .format(aci_const.MODE_INFRA, aci_const.MODE_BAREMETAL, new_mode))
+
+        ctx = context.get_admin_context()
+        curr_mode = self._get_mode(ctx, hostgroup_name)
+        if curr_mode == new_mode:
+            raise web_exc.HTTPBadRequest("{} is already in {} mode".format(hostgroup_name, new_mode))
+
+        # check if there are still  on this hostgroup
+        hg_config = ACI_CONFIG.get_hostgroup(hostgroup_name)
+        if not hg_config:
+            raise web_exc.HTTPBadRequest("Hostgroup {} has no config associated with it".format(hostgroup_name))
+
+        for segment_id in self.db.get_segment_ids_by_physnet(ctx, hg_config['physical_network']):
+            hosts = self.db.get_hosts_on_segment(ctx, segment_id)
+            for host in hg_config['hosts']:
+                if host in hosts:
+                    LOG.error("Denied API request to switch hostgroup %s to %s mode, host %s still bound in segment %s",
+                              hostgroup_name, new_mode, host, segment_id)
+                    raise web_exc.HTTPConflict("Could not switch, there is still a portbinding for host {} "
+                                               "present in segment {}"
+                                               .format(host, segment_id))
+
+        # cleanup old baremetal objects when switching away from baremetal mode
+        baremetal_bindings_cleaned = False
+        if hg_config['hostgroup_mode'] == aci_const.MODE_BAREMETAL:
+            LOG.info("Cleaning baremetal objects for %s from device", hostgroup_name)
+            try:
+                self.rpc_notifier.clean_baremetal_objects(hg_config)
+                baremetal_bindings_cleaned = True
+            except Exception:
+                LOG.exception("Could not clean baremetal bindings from ACI, cleaning will have to be done manually")
+
+        if self.db.set_hostgroup_mode(ctx, hostgroup_name, new_mode):
+            LOG.info("Hostgroup %s set to mode %s", hostgroup_name, new_mode)
+
+            hg_config = ACI_CONFIG.get_hostgroup(hostgroup_name)
+            aci_objects_update_succeeded = False
+            try:
+                # switching policy group of port selectors, etc.
+                self.rpc_notifier.sync_direct_mode_config(hg_config)
+                aci_objects_update_succeeded = True
+            except Exception:
+                LOG.exception("Could not update bindings on ACI, rpc call failed")
+
+            return {
+                "success": True,
+                "baremetal_bindings_cleaned": baremetal_bindings_cleaned,
+                "aci_objects_update_succeeded": aci_objects_update_succeeded,
+            }
+        else:
+            LOG.error("Failed to set hostgroup %s to mode %s - no DB entry?", hostgroup_name, new_mode)
+            raise web_exc.HTTPConflict("Could not set hostmode of hostgroup %s - DB entry could not be found",
+                                       hostgroup_name)
```

Comment:
Going here, we might have cleaned up baremetal objects already, while still staying in `MODE_BAREMETAL` - e.g. if the DB went away or something. Can we catch that case somehow?

Reply:
We could try to do another apicall to reset everything, but for switch from baremetal to infra this would not be problem. For infra to baremetal the host would not be usable until it's synced again. I'm not sure if this is a problem we'll ever run into, but manual intervention would somehow be required.

So... not sure about my options here. I could make the error mesage more explicit, maybe also only for the infra --> baremetal case or I could try to do another API call to the APIC to reset everything. Seems like a lot of work for this though.

Reply:
I've seen code down the line that does an "ensure_baremetal_objects" or something. If that's called every time we try to make a portbinding, then it should still work if we just cleaned and then ended up not switching the mode, right?

Reply:
Everything is synced again on the first deployment to a hostgroup, yes. The baremetal cleanup only removes unused API objects from the APIC.